obspy/io/seisan/sfile.py

```diff
@@ -0,0 +1,269 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+SEISAN SFILE file format support for ObsPy.
+
+:copyright:
+    The ObsPy Development Team (devs@obspy.org)
+:license:
+    GNU Lesser General Public License, Version 3
+    (https://www.gnu.org/copyleft/lesser.html)
+"""
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+from future.builtins import *  # NOQA @UnusedWildImport
+
+import re
+
+from obspy.geodetics import FlinnEngdahl
+
+from obspy import UTCDateTime
+from obspy.core.event import (Catalog, Event, Origin, Magnitude, Pick,
+                              WaveformStreamID, QuantityError)
+
+_fe = FlinnEngdahl()
+
+
+def _get_resource_id(cmtname, res_type, tag=None):
+    """
+    Helper function to create consistent resource ids.
+    """
+    res_id = "smi:local/seisan_sfile/%s/%s" % (cmtname, res_type)
+    if tag is not None:
+        res_id += "#" + tag
+    return res_id
+
+
+def _buffer_proxy(filename_or_buf, function, reset_fp=True,
+                  file_mode="rb", *args, **kwargs):
+    """
+    Calls a function with an open file or file-like object as the first
+    argument. If the file originally was a filename, the file will be
+    opened, otherwise it will just be passed to the underlying function.
+
+    :param filename_or_buf: File to pass.
+    :type filename_or_buf: str, open file, or file-like object.
+    :param function: The function to call.
+    :param reset_fp: If True, the file pointer will be set to the initial
+        position after the function has been called.
+    :type reset_fp: bool
+    :param file_mode: Mode to open file in if necessary.
+    """
+    try:
+        position = filename_or_buf.tell()
+        is_buffer = True
+    except AttributeError:
+        is_buffer = False
+
+    if is_buffer is True:
+        ret_val = function(filename_or_buf, *args, **kwargs)
+        if reset_fp:
+            filename_or_buf.seek(position, 0)
+        return ret_val
+    else:
+        with open(filename_or_buf, file_mode) as fh:
+            return function(fh, *args, **kwargs)
```

Comment:
Thought I've seen this one before.. actually this is the fourth appearance now. Can you do me a favor and move it to core/util? ;-)